profiling/src/timeline.rs

```diff
@@ -0,0 +1,99 @@
+use crate::bindings as zend;
+use crate::PROFILER;
+use crate::REQUEST_LOCALS;
+use std::time::Instant;
+
+/// The engine's original (or previous) `gc_collect_cycles()` function
+static mut PREV_GC_COLLECT_CYCLES: Option<zend::VmGcCollectCyclesFn> = None;
+
+pub fn timeline_minit() {
+    unsafe {
+        // register our function in `gc_collect_cycles`
+        PREV_GC_COLLECT_CYCLES = zend::gc_collect_cycles;
+        zend::gc_collect_cycles = Some(ddog_php_prof_gc_collect_cycles);
+    }
+}
+
+/// Find out the reason for the current garbage collection cycle. If there is
+/// a `gc_collect_cycles` function at the top of the call stack, it is because
+/// of a userland call  to `gc_collect_cycles()`, otherwise the engine decided
+/// to run it.
+unsafe fn gc_reason() -> &'static str {
+    let execute_data = zend::ddog_php_prof_get_current_execute_data();
+    let fname = || execute_data.as_ref()?.func.as_ref()?.name();
+    match fname() {
+        Some(name) if name == b"gc_collect_cycles" => "induced",
+        _ => "engine",
+    }
+}
+
+/// This function gets called whenever PHP does a garbage collection cycle instead of the original
+/// handler. This is done by letting the `zend::gc_collect_cycles` pointer point to this function
+/// and store the previous pointer in `PREV_GC_COLLECT_CYCLES` for later use.
+/// When called, we do collect the time the call to the `PREV_GC_COLLECT_CYCLES` took and report
+/// this to the profiler.
+#[no_mangle]
+unsafe extern "C" fn ddog_php_prof_gc_collect_cycles() -> i32 {
+    if let Some(prev) = PREV_GC_COLLECT_CYCLES {
+        #[cfg(php_gc_status)]
+        let mut status: zend::zend_gc_status = zend::zend_gc_status {
+            #[cfg(php_gc_status_extended)]
+            active: false,
+            #[cfg(php_gc_status_extended)]
+            gc_protected: false,
+            #[cfg(php_gc_status_extended)]
+            full: false,
+            runs: 0,
+            collected: 0,
+            threshold: 0,
+            #[cfg(php_gc_status_extended)]
+            buf_size: 0,
+            num_roots: 0,
+        };
```

Comment:
Florian mentioned using `MaybeUninit` and I think that's better. Just be sure to use `.as_ptr()` instead of using a reference when passing it because taking the reference to an uninitialized struct is undefined behavior in Rust.